compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2405,12 +2406,21 @@ pub struct Constant<'tcx> {
     /// Needed for NLL to impose user-given type constraints.
     pub user_ty: Option<UserTypeAnnotationIndex>,
 
-    pub literal: &'tcx ty::Const<'tcx>,
+    pub literal: ConstantSource<'tcx>,
+}
+
+#[derive(Clone, Copy, PartialEq, PartialOrd, TyEncodable, TyDecodable, Hash, HashStable, Debug)]
+pub enum ConstantSource<'tcx> {
```

Comment:
The enum name is very confusing, since represents a constant *with information about its source*, whereas the name suggests that this is just "what is the source of some constant".

Where is this used? Does it correspond do any of the types in our [previous plan](https://hackmd.io/Qvrj_eOFTkCHZrhJ7f1ItA)?

Reply:
This is an intermediate datastructure to allow us to incrementally move `mir::Constant` to stop using `ty::Const`. I guess we could just name it `mir::ConstKind`?

Reply:
I guess `ConstKind` is consistent with how we usually name the enum in such a struct-enum situation. Or maybe better `ConstantKind` since it is related to `mir::Constant`, not `ty::Const`.